massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router_with_http1 = router_with_http1.add_service(reflection_service);
+            }
+            router_with_http1
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        } else {
+            let mut router = tonic::transport::Server::builder().add_service(svc.clone());
+
+            if config.enable_reflection {
+                let reflection_service = tonic_reflection::server::Builder::configure()
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router = router.add_service(reflection_service);
+            }
+
+            router
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        }
+
+        Ok(StopHandle {
+            server_handler: shutdown_send,
+        })
+    }
+}
+
+/// Used to be able to stop the gRPC API
+pub struct StopHandle {
+    server_handler: oneshot::Sender<()>,
+}
+
+impl StopHandle {
+    /// stop the gRPC API gracefully
+    pub fn stop(self) {
+        match self.server_handler.send(()) {
+            Ok(_) => {
+                info!("gRPC API finished cleanly");
+            }
+            Err(err) => warn!("gRPC API thread panicked: {:?}", err),
+        }
+    }
+}
+
+fn match_for_io_error(err_status: &tonic::Status) -> Option<&std::io::Error> {
+    let mut err: &(dyn Error + 'static) = err_status;
+
+    loop {
+        if let Some(io_err) = err.downcast_ref::<std::io::Error>() {
+            return Some(io_err);
+        }
+
+        // h2::Error do not expose std::io::Error with `source()`
+        // https://github.com/hyperium/h2/pull/462
+        if let Some(h2_err) = err.downcast_ref::<h2::Error>() {
+            if let Some(io_err) = h2_err.get_io() {
+                return Some(io_err);
+            }
+        }
+
+        err = match err.source() {
+            Some(err) => err,
+            None => return None,
+        };
+    }
+}
+
+#[tonic::async_trait]
+impl grpc::grpc_server::Grpc for MassaService {
+    async fn get_selector_draws(
+        &self,
+        request: tonic::Request<grpc::GetSelectorDrawsRequest>,
+    ) -> Result<tonic::Response<grpc::GetSelectorDrawsResponse>, tonic::Status> {
+        let selector_controller = self.selector_controller.clone();
+        let config = self.grpc_config.clone();
+        let inner_req = request.into_inner();
+        let id = inner_req.id.clone();
+
+        let addresses_res = inner_req
+            .queries
+            .into_iter()
+            .map(|query| {
+                //TODO remove unwrap ?
+                Address::from_str(query.filter.unwrap().address.as_str())
+            })
+            .collect::<Result<Vec<_>, _>>();
+
+        match addresses_res {
```

Comment:
if match has Ok or Err wich one of them is long i prefer put the more short way in first like 'early return' pattern : 

```rust
match addresses_res {
     Err(e) => return .... , 
     Ok(addresses) = > {
        // long code
        // ....
     }
}
```

but it doesn't matter

Reply:
I'm a partisan of `early return` pattern but changing the match arms order will destabilize Rust developers

Reply:
don't be afraid to mess them up :smiling_imp: